DOCS-7826-Sockets-examples-editorial-kt (#717)
mulesoft/docs-connectors: modules/ROOT/pages/edits/sockets-connector-examples.adoc

= Sockets Examples - Mule 4
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]
:imagesdir: ../../assets/images/

> Anypoint Connector for Sockets (Sockets Connector) enables you to send or receive transport layer packages through both UDP and TCP sockets. The following examples help you to use the available operations considering the Transport Layer protocol you want to implement in your application.

Reviewer: ```suggestion
Anypoint Connector for Sockets (Sockets Connector) enables you to send or receive Transport Layer packages through both UDP and TCP sockets. The following examples help you to implement these packages in your application.
```
Does this suggestion hinder your intended meaning? 
The troublesome phrase is "...options considering the TPL protocol you want to implement...," specifically the word "considering."
Perhaps you mean 
```suggestion
Anypoint Connector for Sockets (Sockets Connector) enables you to send or receive Transport Layer packages through both UDP and TCP sockets. The following examples help you to implement the protocol as required for your application.
```

> == Send a Message

Reviewer: Example of Sending a Message and Testing the Outcome?

> In the following example, you have set up your Socket Listener and you want to build an application that upon receiving specific data from an external client, performs a database query and sends a message to a UDP Server Socket with the obtained results. Note that, even though we chose UDP for this example, you can use the Send operation to send messages to both TCP and UDP sockets:

Reviewer: I briefly looked up socket terminology and did not see much capitalization, so I am guessing that we want to use "socket Listener" or "Listener socket" and "UDP server socket". Please tell me what you think is correct and I will add these to our Term Presentation list.

```suggestion
In the following example, you have set up your socket Listener and you want to build an application that, upon receiving specific data from an external client, performs a database query and sends the results to a UDP server socket. Note that although this example uses UDP, you can send messages to both TCP and UDP sockets:
```

		</sockets:tcp-listener-connection>
	</sockets:listener-config>

	<db:config name="Database_Config" doc:name="Database Config">
		<db:oracle-connection host="localhost" user="sys as sysdba" password="Oradoc_db1" serviceName="orclpdb1.localdomain" />
	</db:config>

	<sockets:request-config name="Sockets_Request_config" doc:name="Sockets Request config" >
		<sockets:udp-requester-connection host="localhost" port="8085" />
	</sockets:request-config>

	<flow name="socketSearchDatabaseAndReturnResult">
		<sockets:listener doc:name="Listener" config-ref="Sockets_Listener_config" outputMimeType="application/json"/>
		<db:select doc:name="Select" config-ref="Database_Config">
			<db:sql>#["SELECT first_name, last_name from SYSTEM.MUSICIANS where id = :id"]</db:sql>
			<db:input-parameters><![CDATA[#[{ 'id': payload.id }]]]></db:input-parameters>
		</db:select>
		<logger level="INFO" doc:name="Logger" message="#[payload]" />
		<sockets:send doc:name="Send" config-ref="Sockets_Request_config">
			<sockets:content ><![CDATA[#[output application/json --- 'name': payload[0].FIRST_NAME as String ++ ' ' ++ payload[0].LAST_NAME as String]]]></sockets:content>
		</sockets:send>
	</flow>
----

> To test this example:

Reviewer: When we present extended syntax for an example, we don't typically follow that with "testing," so I am wondering what this is or why the audience would need to perform it (assuming that the example is used to show proper syntax input). 
Is what follows really the example, but with real or actual customer values?
Is this a test of implementation or results?

> . Initialize an Oracle database using the following script:

Reviewer: ```suggestion
. Initialize an Oracle database by using the following script:
```

[source,xml,linenums]
----
    CREATE TABLE SYSTEM.MUSICIANS(
        id INTEGER GENERATED BY DEFAULT AS IDENTITY,
        age INTEGER,
        first_name VARCHAR2(100),
        last_name VARCHAR2(100),
        PRIMARY KEY(id)
    );

    INSERT INTO SYSTEM.MUSICIANS(age, first_name, last_name) VALUES(37, 'Farrokh', 'Bulsara');
    INSERT INTO SYSTEM.MUSICIANS(age, first_name, last_name) VALUES(36, 'Brian Harold', 'May');
    INSERT INTO SYSTEM.MUSICIANS(age, first_name, last_name) VALUES(35, 'Roger', 'Meddows Taylor');
    INSERT INTO SYSTEM.MUSICIANS(age, first_name, last_name) VALUES(32, 'John', 'Deacon');
----
[start=2]

> . Start a UDP server at port 8085, to which the results will be sent. Do this by running netcat command `nc -ul 8085`.

Reviewer: ```suggestion
. At port 8085, start a UDP server to which the results will be sent by running the netcat command `nc -ul 8085`:
```
I wonder if we should include the part about the netcat command, which that is not shown in the following image? If this is a later action, maybe it should have its own step below?

[start=3]

> . Run your Mule application and start a TCP Client Socket to send the expected query parameter. Do this by running netcat command `nc localhost 8082` and sending the following data `{"id": 2}`.

Reviewer: Ditto re. capitalization; let me know.
```suggestion
. Run your Mule application and start a TCP client socket to send the expected query parameter by running netcat command `nc localhost 8082` and sending the data `{"id": 2}`.
```
An alternative is to end the original first sentence with a colon and present the command and data using an example syntax block.

== Send and Receive Messages

> Because Sockets are full-duplex, they allow bidirectional communication, thus you can send and receive messages simultaneously. The following example operation is slightly different from the previous one because it will first send a message and then expect a response. +

Reviewer: When I read further, I noted some repetition, so I suggest merging the two related texts into this one. Note that I reworded "the previous one" to ensure reusability by naming the text instead of giving it a position.

```suggestion
Because sockets are full-duplex, they allow bidirectional communication, enabling you to send and receive messages simultaneously. The following example operation is slightly different from the example of only sending a message because this example not only sends data to the client but also blocks operations until it receives a response or times out (null payload). +
```

This operation can be used for both TCP and UDP sockets.

The operation will:

. Send the data using the client associated to the RequesterConnection.

> . Block until data is received or Timeout is met (in which case a null payload will be returned).

Reviewer: I think that this information should be merged into the introduction, because the intro already specifies sending data and then waiting for a response, and is then followed by the bit about TCP and UDP, and then here returns to the sending and waiting. It's a bit circular and repetitive.

> Using the previous example, now you add a new flow `socketSendQueryAndReceiveResult` that connects to the Listener from the `socketSearchDatabaseAndReturnResult` flow, send the query string and wait for the response with the query result. Once the result has been received, it will be written to a file:

Reviewer: ```suggestion
Building on the example of only sending data, in this scenario you add a new flow, `socketSendQueryAndReceiveResult`, that connects to the Listener from the `socketSearchDatabaseAndReturnResult` flow, send the query string, and wait for the response, which is then written to a file:
```

		<file:connection />
	</file:config>

	<flow name="socket-testFlow">
		<sockets:listener doc:name="Listener" config-ref="Sockets_Listener_config" outputMimeType="application/json" />
		<db:select doc:name="Select" config-ref="Database_Config">
			<db:sql >#["SELECT first_name, last_name from SYSTEM.MUSICIANS where id = :id"]</db:sql>
			<db:input-parameters><![CDATA[#[{ 'id': payload.id }]]]></db:input-parameters>
		</db:select>
		<ee:transform doc:name="Transform Message">
			<ee:message >
				<ee:set-payload><![CDATA[%dw 2.0
					output application/json --- 'name': payload[0].FIRST_NAME as String ++ ' ' ++ payload[0].LAST_NAME as String]]>
				</ee:set-payload>
			</ee:message>
		</ee:transform>
	</flow>

	<flow name="socketSendAndReceive" >
		<sockets:listener doc:name="Listener" config-ref="Sockets_Listener_config1" outputMimeType="application/json"/>
		<sockets:send-and-receive doc:name="Send and receive" config-ref="Sockets_Request_config" outputMimeType="application/json">
		</sockets:send-and-receive>
		<logger level="INFO" doc:name="Logger" message='#[payload]'/>
		<file:write path="musicians.json" config-ref="File_Config" mode="CREATE_NEW" />
	</flow>
----

> To test this example, use the same database setup from the previous example, run the mule application and hit the endpoint using the following netcat command: `nc localhost 8085` and send the following data `{"id":2}`.

Reviewer: ```suggestion
To test this example, use the same database setup as used for the sending example, run the Mule application, access the endpoint using the following netcat command `nc localhost 8085`, and send the data `{"id":2}`.
```
Think about whether these examples will continue to "travel" together. It is kind of risky to base one block of text on that of another.
If you do choose to rely on the previous example, are these the same testing steps, and can you say simply "You can test this example using the same steps you used to test the sending-only example"?
